pkg/controller/nodedaemon/daemonsets.go

```diff
@@ -0,0 +1,254 @@
+package nodedaemon
+
+import (
+	"context"
+
+	"github.com/openshift/local-storage-operator/pkg/controller/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// createOrUpdateDaemonset fetches and creates/updates ds according to desiredDs
+// fields that are not allowed update at runtime are not copied unless the creationtimestamp is zero
+func createOrUpdateDaemonset(c client.Client, desiredDs *appsv1.DaemonSet) (controllerutil.OperationResult, error) {
+	ds := &appsv1.DaemonSet{ObjectMeta: metav1.ObjectMeta{Name: desiredDs.GetName(), Namespace: desiredDs.GetNamespace()}}
+	return controllerutil.CreateOrUpdate(context.TODO(), c, ds, func() error {
+		updateDaemonSet(*desiredDs, ds)
+		return nil
+	})
+}
+
+// updateDaemonSet copies the daemonset from in to out
+// fields that are not allowed update at runtime are not copied unless the creationtimestamp is zero
+func updateDaemonSet(in appsv1.DaemonSet, out *appsv1.DaemonSet) {
+
+	if out.CreationTimestamp.IsZero() {
+		in.DeepCopyInto(out)
+		return
+	}
+	out.ObjectMeta.Labels = in.GetLabels()
+	out.ObjectMeta.Annotations = in.ObjectMeta.Annotations
+	out.ObjectMeta.OwnerReferences = in.ObjectMeta.OwnerReferences
+
+	out.Spec.Template.ObjectMeta.Annotations = in.Spec.Template.ObjectMeta.Annotations
+	out.Spec.Template.Spec = in.Spec.Template.Spec
+
+	out.Spec.UpdateStrategy = in.Spec.UpdateStrategy
+	out.Spec.MinReadySeconds = in.Spec.MinReadySeconds
+	out.Spec.RevisionHistoryLimit = in.Spec.RevisionHistoryLimit
+	return
```

Comment:
copying @jsafrane 's [comment](https://github.com/openshift/local-storage-operator/pull/123#discussion_r456509827) here as this is where the change is. #123 is based on this.



I checked the way you apply DaemonSet and I can immediately spot an issue: what if user (or another operator / controller / whatever) changes Spec.Selector? This code does not fix the selector back to the state the operator wants it. This may result in unexpected and very hard to debug behavior of the whole LSO.

Not mentioning that you're abusing the API server by calling Update even when there may be no actual update necessary.

Of course, you can add the selector to things the operator replaces. But explicitly listing fields you're going to fix is very brittle. In the future, Kubernetes may introduce a new DaemonSet.Spec field that the operator does not know about and it will not fix it. And it may break the operator again.

What is 100% safe and future proof is to remember the last Generation of the DaemonSet that LSO created / updated. When the operator gets an event that the DaemonSet has updated, it can compare the current Generation with the last correct Generation and immediately see that someone / something has changed DaemonSet.Spec and overwrite the whole Spec. And that's why we're tracking generations of children in LSO Status.

Reply:
> I checked the way you apply DaemonSet and I can immediately spot an issue: what if user (or another operator / controller / whatever) changes Spec.Selector? This code does not fix the selector back to the state the operator wants it. This may result in unexpected and very hard to debug behavior of the whole LSO.

`DaemonSetSpec.Spec.Selector` is an immutable field.

> Not mentioning that you're abusing the API server by calling Update even when there may be no actual update necessary.

[Not true](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L216):
```
	if equality.Semantic.DeepEqual(existing, obj) {
		return OperationResultNone, nil
	}
```
> In the future, Kubernetes may introduce a new DaemonSet.Spec field that the operator does not know about and it will not fix it. And it may break the operator again.

we'd have to implement using that field in code too.

> What is 100% safe and future proof is to remember the last Generation of the DaemonSet that LSO created / updated. When the operator gets an event that the DaemonSet has updated, it can compare the current Generation with the last correct Generation and immediately see that someone / something has changed DaemonSet.Spec and overwrite the whole Spec. And that's why we're tracking generations of children in LSO Status.

Here,  that is taken care of by the watch mechanism. It ensures that the reconciler runs every time it starts or sees a cache update (by tracking the generation).

Reply:
> if equality.Semantic.DeepEqual(existing, obj) {
>		return OperationResultNone, nil
>	}

If you replace `obj.Spec.Template`, then it will always return that the objects are different because of defaulting in API server.

> Here, that is taken care of by the watch mechanism. It ensures that the reconciler runs every time it starts or sees a cache update (by tracking the generation).

I am not sure I understand. Where LSO tracks generation of the newly introduced DaemonSet?

Reply:
> If you replace obj.Spec.Template, then it will always return that the objects are different because of defaulting in API server.

Ah, I see, I'll investigate. Does apply not also hit the API server?

 > I am not sure I understand. Where LSO tracks generation of the newly introduced DaemonSet?

It doesn't but the reconciler runs for every generation change, but I see what you mean. I will look into implementing generation tracking for the purpose of bubbling up errors.

Reply:
Yes, `ApplyDaemonSet` in library-go hits API server, and it mostly does Get, not Update, which is not that invasive.

Reply:
Solved by defining the daemonset mutateFuncs to not override defaults.